### PR TITLE
External api integration

### PR DESCRIPTION
**Pull request by alanlamaswolox**
## Summary
Created OpenLibraryService service for connect Open Library API and return specific books

## Trello
https://trello.com/c/yUjfhjGn/38-28-integraci%C3%B3n-api-externa

### CODE REVIEW

File: app/controllers/api/v1/books_controller.rb [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require_relative '../../../services/open_library_service'
```
Comment (nicozar95) [inline app/controllers/api/v1/books_controller.rb]: This is not necesary

File: app/services/open_library_service.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+class OpenLibraryService
+  def book_info(isbn)
+    url = "https://openlibrary.org/api/books?format=json&jscmd=data&bibkeys=ISBN:#{isbn}"
```
Comment (nicozar95) [inline app/services/open_library_service.rb]: Move this to base uri of Httparty

File: app/services/open_library_service.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    ol_response = HTTParty.get(url)["ISBN:#{isbn}"]
+    hash_response = {}
```
Comment (nicozar95) [inline app/services/open_library_service.rb]: Make all this into a private method (line 7 to 15) It's hard to read

File: config/routes.rb [hunk 4]
```diff
@@ -5,6 +5,8 @@
 
   api_version(module: 'api/v1', path: { value: 'api/v1' }, defaults: { format: :json }) do
     resources :books, only: [:index,:show]
+    get '/books/open_library/:isbn', to: 'books#open_library'
```
Comment (nicozar95) [inline config/routes.rb]: I don't think :isbn is necesary here

**Review by nicozar95 — CHANGES_REQUESTED** [review]
Change all this and I will take a look to it

File: app/services/open_library_service.rb (hunk 2 above)
Comment (alanlamaswolox) [reply app/services/open_library_service.rb]: Excuse me, Nico, i don't understand this

Comment (nicozar95) [reply app/services/open_library_service.rb]: https://www.rubydoc.info/github/jnunemaker/httparty/HTTParty%2FClassMethods:base_uri

File: app/controllers/api/v1/books_controller.rb [hunk 5]
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+# require_relative '../../../services/open_library_service'
```
Comment (nicozar95) [inline app/controllers/api/v1/books_controller.rb]: Remove comment

File: app/services/open_library_service.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  private
+
+  def build_hash(isbn)
+    url_isbn = "ISBN:#{isbn}"
+    @options = { query: { bibkeys: url_isbn, format: 'json', jscmd: 'data' } }
+    ol_response = self.class.get('/api/books', @options)[url_isbn]
+    hash_response = {}
+    hash_response['isbn'] = isbn
```
Comment (nicozar95) [inline app/services/open_library_service.rb]:
```
def hash_response
{
  isbn: ..,
  title: ..,
  . . .
end
```

Move the hash_response into a new method

**Review by nicozar95 — CHANGES_REQUESTED** [review]
Almost

File: app/services/open_library_service.rb (hunk 6 above)
Comment (nicozar95) [reply app/services/open_library_service.rb]: This is still not fix

File: app/services/open_library_service.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class OpenLibraryService
+  include HTTParty
+  base_uri 'https://openlibrary.org'
+  def book_info(isbn)
+    hash_response = return_hash(isbn)
```
Comment (nicozar95) [inline app/services/open_library_service.rb]: Change the method name to `build_hash(isbn)`

**Review by nicozar95 — CHANGES_REQUESTED** [review]
Please fix the hash build

**Review by fedeest — CHANGES_REQUESTED** [review]
Make some tests for your service and for your endpoint.

Also, you should manage error cases for every external comunication of your API. In this case, you should check the case where you can't connect to the external service.

In normal cases, the response from the other service could be something different than 200. In those cases you should check that too. In this case, 200 is always the response, so just add the necessary code to manage that case, and check in a test that you are managing that error case.

File: app/services/open_library_service.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  def isbn_hash(isbn)
+    hash_response = book_information(isbn)
+    hash_response
+  end
+
+  private
+
+  def book_information(isbn)
+    url_isbn = "ISBN:#{isbn}"
+    @options = { query: { bibkeys: url_isbn, format: 'json', jscmd: 'data' },
```
Comment (nicozar95) [inline app/services/open_library_service.rb]: Move the query option to another method

File: app/services/open_library_service.rb [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+                 headers: { 'Accept' => 'application/json' } }
```
Comment (nicozar95) [inline app/services/open_library_service.rb]: Move the header to another method

File: app/services/open_library_service.rb [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    begin
+      ol_response = self.class.get('/api/books', @options)[url_isbn]
+    rescue SocketError
+      return 'The URI is wrong, change it please'
```
Comment (nicozar95) [inline app/services/open_library_service.rb:13]: Wait... What? why are you catching a SocketError?

File: app/services/open_library_service.rb [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    end
+    hash = build_hash(ol_response, isbn)
+    hash
+  end
+
+  def build_hash(ol_response, isbn)
```
Comment (nicozar95) [inline app/services/open_library_service.rb]: 👍

File: app/services/open_library_service.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+class OpenLibraryService
+  include HTTParty
+  base_uri 'https://openlibrary.org'
```
Comment (nicozar95) [inline app/services/open_library_service.rb]: Move the string 'https://openlibrary.org' to a constant

File: spec/services/open_library_service.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+  describe 'GET #isbn_hash' do
+    let!(:service) { OpenLibraryService.new }
+    let(:hash) { service.isbn_hash('0385472579') }
+    context 'when service connects successfully with external API' do
+      it "its response's title is Zen speaks" do
+        expect(hash[:title]).to eq 'Zen speaks'
+      end
+    end
+
+    context "When service fails to connect to external API's URI" do
```
Comment (nicozar95) [inline spec/services/open_library_service.rb:44]: I don't think this test is necesary

**Review by nicozar95 — CHANGES_REQUESTED** [review]
It's okay, if you want to test external API you should mock the response with a helper like WebMock. Why? Because it's dangerous to test external API to the real API, in this case you are searching if a book exist or not in OpenLibrary, but maybe you will have to test more sensitive data, like credits card for example and you MUST MOCK THOSE.

How you are supposed to mock it? Get a response of the success and the failes API (sucess, for example you have to ask for a book that exist. faild, when a book doesnt exist) and create fixtures with those which WebMock will use as a body.

File: app/services/open_library_service.rb [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+  def isbn_hash(isbn)
+    hash_response = book_information(isbn)
+    hash_response
```
Comment (fedeest) [inline app/services/open_library_service.rb]: Unnecesary variable assignation. Weird rubocop didn't complain about this

File: app/services/open_library_service.rb (hunk 10 above)
Comment (fedeest) [reply app/services/open_library_service.rb:13]: This is the case in which the API is non respondant. HTTParty throws this error.

File: app/services/open_library_service.rb [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    url_isbn = "ISBN:#{isbn}"
+    @options = { query: { bibkeys: url_isbn, format: 'json', jscmd: 'data' },
+                 headers: { 'Accept' => 'application/json' } }
+    begin
+      ol_response = self.class.get('/api/books', @options)[url_isbn]
+    rescue SocketError
+      return 'The URI is wrong, change it please'
+    end
+    hash = build_hash(ol_response, isbn)
+    hash
```
Comment (fedeest) [inline app/services/open_library_service.rb]: Again, useless variable assignation, just return the value

File: spec/services/open_library_service.rb (hunk 13 above)
Comment (fedeest) [reply spec/services/open_library_service.rb:44]: It is!

File: spec/services/open_library_service.rb [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+      it 'raises and SocketError exception' do
+        allow(OpenLibraryService).to receive(:get).and_raise(SocketError)
+        response = service.isbn_hash('0385472579')
```
Comment (fedeest) [inline spec/services/open_library_service.rb]: This kind of things should be placed outside the it, but inside this context.

The it should only have the expect. You can place those in a before block

File: app/controllers/api/v1/open_library_controller.rb [hunk 17]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+module Api
+  module V1
+    class OpenLibraryController < ApiController
+      def show
+        ol_response = OpenLibraryService.new.isbn_hash(params[:id])
+        render json: ol_response
```
Comment (fedeest) [inline app/controllers/api/v1/open_library_controller.rb]: Again, useless variable assignation

**Review by fedeest — CHANGES_REQUESTED** [review]

File: app/services/open_library_service.rb [hunk 18]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+class OpenLibraryService
+  include HTTParty
+  HTTPARTY_BASE_URL = 'https://openlibrary.org'
+  base_uri HTTPARTY_BASE_URL
+
+  def isbn_hash(isbn)
+    book_information(isbn)
```
Comment (fedeest) [inline app/services/open_library_service.rb]: Don't do wrapper methods. This method does nothing but call the other method.

File: app/services/open_library_service.rb [hunk 19]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+class OpenLibraryService
+  include HTTParty
```
Comment (fedeest) [inline app/services/open_library_service.rb:4]: .freeze

File: app/services/open_library_service.rb [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+      title: ol_response['title'],
+      subtitle: ol_response['subtitle'],
+      number_of_pages: ol_response['number_of_pages'],
+      authors: ol_response['authors']
+    }
+  end
+
+  def build_query_options(url_isbn)
+    { query: { bibkeys: url_isbn, format: 'json', jscmd: 'data' } }
+  end
```
Comment (fedeest) [inline app/services/open_library_service.rb:44]: I think you are missing the headers

**Review by fedeest — CHANGES_REQUESTED** [review]
Remember to also mock the requests on your tests

File: app/services/open_library_service.rb [hunk 21]
```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+class OpenLibraryService
+  include HTTParty.freeze
```
Comment (nicozar95) [inline app/services/open_library_service.rb]: lol the .freeze is for the constant -> https://blog.honeybadger.io/when-to-use-freeze-and-frozen-in-ruby/

File: app/services/open_library_service.rb [hunk 22]
```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+class OpenLibraryService
+  include HTTParty
+  HTTPARTY_BASE_URL = 'https://openlibrary.org' # tira quilombos con rubocop
```
Comment (fedeest) [inline app/services/open_library_service.rb]: Don't leave this kind of comments in the code.

File: app/services/open_library_service.rb [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+  base_uri HTTPARTY_BASE_URL
+
+  def book_hash(isbn)
```
Comment (fedeest) [inline app/services/open_library_service.rb]: This name for the method if kind non descriptive.

You should use something like `get_books`, describing the action to be done in the request.

**Comment by fedeest** [comment]
Lol this is still here.

**Pull request closed without merging**